modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -4683,3 +4688,150 @@ def compare(self, other, **kwargs):
                 other._modin_frame,
             )
         )
+
+    def case_when(self, caselist):
+        if Engine.get() in ("Unidist", "Dask"):
+            # https://github.com/modin-project/modin/issues/6973
+            return super().case_when(caselist)
+
+        frame = self._modin_frame
+        # If all the conditions are callable and the replacements are either
+        # callable or scalar, use map().
+        if all(
+            callable(c[0]) and (callable(c[1]) or not is_list_like(c[1]))
+            for c in caselist
+        ):
+            frame = frame.map(func=_case_when, func_args=[caselist], lazy=True)
+            return self.__constructor__(frame)
+
+        parts = [p[0] for p in frame._partitions]
+        parts_len = len(parts)
+        lengths = frame._get_lengths(parts, Axis.ROW_WISE)
+        new_caselist = []
+        is_trivial_idx = None
+        for c in caselist:
+            c = list(c)
+            for i in range(2):
+                data = c[i]
+                if not isinstance(data, pd.Series):
```

Comment:
We shouldn't operate on the high-layer objects in the lower layers? See a comment above.